File: ACP_comm_test/ACP/acpcomms/python3/messenger.py
"""
Messenger - ZeroMQ messaging wrapper with internal BSON serialization.
Publisher/subscriber scripts do NOT need bson, json, or zmq imports.
"""

import zmq
import threading
from typing import Dict, Any, Callable, Optional

# Keep your original package structure
from .acp_comms import ACPComms

# Internal serialization
from bson import dumps as bson_dumps, loads as bson_loads
import json


class Messenger(ACPComms):
    """ZeroMQ-based messaging implementation with automatic BSON serialization/deserialization."""

    def __init__(self):
        super().__init__()
        self.context: Optional[zmq.Context] = None
        self.socket: Optional[zmq.Socket] = None
        self.endpoint: str = ""
        self.socketType: int = zmq.SUB
        self.listenerThread: Optional[threading.Thread] = None
        self.listening: bool = False
        self.messageHandler: Optional[Callable[[dict], None]] = None

        # ------------------------------------------------------------------
        # Default JSON message template
        # Industry-standard: centrally defined to avoid duplication
        # Publishers may override fields when sending messages
        # ------------------------------------------------------------------
        self.messageTemplate: Dict[str, Any] = {
            "sensor_id": None,
            "sensor_type": None,
            "data_type": None,
            "timestamp": None,
            "location": {
                "lat": None,
                "lon": None
            },
            "metadata": {
                "units": None,
                "schema_version": "1.0",
                "tags": []
            },
            "data": {}
        }

    # -----------------------------
    # Configure ZMQ socket
    # -----------------------------
    def configure(self, configuration: Dict[str, Any]) -> None:
        self.config.update(configuration)

        self.endpoint = self.config.get("endpoint", "tcp://localhost:5555")
        socket_type_str = self.config.get("socketType", "SUB")

        socket_type_map = {
            "PUB": zmq.PUB,
            "SUB": zmq.SUB,
            "REQ": zmq.REQ,
            "REP": zmq.REP,
            "PUSH": zmq.PUSH,
            "PULL": zmq.PULL
        }

        if socket_type_str not in socket_type_map:
            raise ValueError(f"Invalid socket type: {socket_type_str}")

        self.socketType = socket_type_map[socket_type_str]

    # -----------------------------
    # Connect / Disconnect
    # -----------------------------
    def connect(self) -> None:
        if self.connected:
            raise RuntimeError("Already connected")

        self.context = zmq.Context()
        self.socket = self.context.socket(self.socketType)

        if self.config.get("bind", False):
            self.socket.bind(self.endpoint)
        else:
            self.socket.connect(self.endpoint)

        if self.socketType == zmq.SUB:
            topic = self.config.get("topic", "")
            self.socket.subscribe(topic.encode())

        self.connected = True

    def disconnect(self) -> None:
        if not self.connected:
            return

        self.stopListener()

        if self.socket:
            self.socket.close()
        if self.context:
            self.context.term()

        self.connected = False

    # -----------------------------
    # Internal serialization helpers
    # -----------------------------
    def serialize(self, obj: dict) -> bytes:
        """dict → BSON bytes"""
        return bson_dumps(obj)

    def deserialize(self, data: bytes) -> dict:
        """BSON bytes → dict"""
        return bson_loads(data)

    def deserialize_to_json(self, data: bytes) -> str:
        """BSON bytes → JSON string"""
        return json.dumps(self.deserialize(data))

    # -----------------------------
    # Message Template
    # -----------------------------
    def setTemplate(self, jsonTemplate: Dict[str, Any]) -> None:
        # Industry-standard: allow template replacement if needed
        self.messageTemplate = jsonTemplate

    def getTemplate(self) -> Dict[str, Any]:
        """Return a copy of the current template for publishers."""
        return self.messageTemplate.copy()

    def sendMessageTemplate(self, **kwargs):
        if self.messageTemplate is None:
            raise ValueError("Message template not set")

        # Industry-standard: merge template defaults with overrides
        populated = json.loads(json.dumps(self.messageTemplate))  # deep copy
        populated.update(kwargs)

        bson_bytes = self.serialize(populated)
        self.sendMessage(bson_bytes)

    # -----------------------------
    # Raw send/receive
    # -----------------------------
    def sendMessage(self, message: bytes) -> None:
        if not self.connected:
            raise RuntimeError("Not connected")
        self.socket.send(message)

    def receiveMessage(self) -> bytes:
        if not self.connected:
            raise RuntimeError("Not connected")
        return self.socket.recv()

    def receiveMessageObject(self) -> dict:
        return self.deserialize(self.receiveMessage())

    # -----------------------------
    # Background listener
    # -----------------------------
    def setMessageHandler(self, handler: Callable[[dict], None]) -> None:
        """Handler receives **Python dict**, not raw bytes."""
        self.messageHandler = handler

    def startListener(self) -> None:
        if not self.connected:
            raise RuntimeError("Not connected")
        if self.listening:
            raise RuntimeError("Listener already running")

        self.listening = True

        def loop():
            while self.listening:
                try:
                    raw = self.socket.recv()

                    json_string = self.deserialize_to_json(raw)

                    if self.messageHandler:
                        self.messageHandler(json_string)

                except Exception as e:
                    if self.listening:
                        print("Error receiving:", e)

        self.listenerThread = threading.Thread(target=loop, daemon=True)
        self.listenerThread.start()

    def stopListener(self) -> None:
        self.listening = False
        if self.listenerThread:
            self.listenerThread.join(timeout=1.0)

File: ACP_comm_test/ACP/acpcomms/java/Messenger.java
package ACP.acpcomms.java;

import org.zeromq.ZMQ;
import org.zeromq.ZContext;
import java.util.Map;
import java.util.HashMap;
import java.util.function.Consumer;

import com.fasterxml.jackson.databind.ObjectMapper;

public class Messenger extends ACPComms {
    private ZContext context;
    private ZMQ.Socket socket;
    private String endpoint;
    private int socketType;
    private Thread listenerThread;
    private volatile boolean listening;
    private Consumer<byte[]> messageHandler;

    // -----------------------------
    // JSON TEMPLATE SUPPORT (ADDED)
    // -----------------------------
    private Map<String, Object> messageTemplate;
    private final ObjectMapper jsonMapper = new ObjectMapper();


    public Messenger() {
        super();
    }

    @Override
    public void configure(Map<String, Object> configuration) throws Exception {
        this.config.putAll(configuration);

        // TODO: need to add configuration type for other protocols
        // this.protocol = (String) config.getOrDefault("protocol", "ZMQ");

        this.endpoint = (String) config.getOrDefault("endpoint", "tcp://localhost:5555");
        String type = (String) config.getOrDefault("socketType", "SUB"); // Only applies to MQ need to make submodules for supported protocols

        switch (type) {
            case "PUB" -> this.socketType = ZMQ.PUB;
            case "SUB" -> this.socketType = ZMQ.SUB;
            case "REQ" -> this.socketType = ZMQ.REQ;
            case "REP" -> this.socketType = ZMQ.REP;
            case "PUSH" -> this.socketType = ZMQ.PUSH;
            case "PULL" -> this.socketType = ZMQ.PULL;
            default -> throw new IllegalArgumentException("Invalid socket type: " + type);
        }
    }


    @Override
    public void connect() throws Exception {
        if (connected) {
            throw new IllegalStateException("Already connected");
        }

        context = new ZContext();
        socket = context.createSocket(socketType);

        boolean bind = (boolean) config.getOrDefault("bind", false);

        if (bind) {
            socket.bind(endpoint);
        } else {
            socket.connect(endpoint);
        }

        // For SUB sockets, subscribe to topics
        if (socketType == ZMQ.SUB) {
            String topic = (String) config.getOrDefault("topic", ""); // TODO: parameterize topics
            socket.subscribe(topic.getBytes());
        }

        connected = true;
    }


    @Override
    public void disconnect() throws Exception {
        if (!connected) {
            return;
        }

        stopListener();

        if (socket != null) {
            socket.close();
        }
        if (context != null) {
            context.close();
        }

        connected = false;
    }


    // -----------------------------
    // JSON TEMPLATE METHODS (ADDED)
    // -----------------------------

    /** Store a JSON message template. */
    public void setTemplate(Map<String, Object> template) {
        this.messageTemplate = new HashMap<>(template);
    }

    /** Return a copy of the stored template (to avoid mutation). */
    public Map<String, Object> getTemplate() {
        if (this.messageTemplate == null) {
            return null;
        }
        return new HashMap<>(this.messageTemplate);
    }

    /**
     * Populate the JSON template with caller-supplied overrides and send.
     * Equivalent to Python's sendMessageTemplate(temp=42, mode="run").
     */
    public void sendMessageTemplate(Map<String, Object> overrides) throws Exception {
        if (messageTemplate == null) {
            throw new IllegalStateException("Message template is not set");
        }

        // Copy template so we don’t mutate the original
        Map<String, Object> merged = new HashMap<>(messageTemplate);
        merged.putAll(overrides);

        String jsonString = jsonMapper.writeValueAsString(merged);
        sendMessage(jsonString);
    }

    // -----------------------------
    // ZMQ SEND / RECEIVE METHODS
    // -----------------------------

    public void sendMessage(byte[] message) throws Exception {
        if (!connected) {
            throw new IllegalStateException("Not connected");
        }
        socket.send(message, 0);
    }


    public void sendMessage(String message) throws Exception {
        sendMessage(message.getBytes());
    }


    public byte[] receiveMessage() throws Exception {
        if (!connected) {
            throw new IllegalStateException("Not connected");
        }
        return socket.recv(0);
    }


    public String receiveMessageString() throws Exception {
        return new String(receiveMessage());
    }


    public void setMessageHandler(Consumer<byte[]> handler) {
        this.messageHandler = handler;
    }


    public void startListener() throws Exception {
        if (!connected) {
            throw new IllegalStateException("Not connected");
        }
        if (listening) {
            throw new IllegalStateException("Listener already running");
        }

        listening = true;

        listenerThread = new Thread(() -> {
            while (listening && !Thread.currentThread().isInterrupted()) {
                try {
                    byte[] message = socket.recv(0);
                    if (message != null && messageHandler != null) {
                        messageHandler.accept(message);
                    }
                } catch (Exception e) {
                    if (listening) {
                        System.err.println("Error receiving message: " + e.getMessage());
                    }
                }
            }
        });

        listenerThread.start();
    }


    public void stopListener() {
        listening = false;
        if (listenerThread != null) {
            listenerThread.interrupt();
            try {
                listenerThread.join(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}
